logger/logger.go

```diff
@@ -0,0 +1,118 @@
+package logger
+
+import (
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var (
+	ErrEmptyService = errors.New("empty service name")
+
+	mu sync.Mutex
+
+	defaultCfg = zap.Config{
+		Encoding:          "console",
+		Level:             zap.NewAtomicLevelAt(zapcore.Level(zapcore.DebugLevel)),
+		OutputPaths:       []string{"stdout"},
+		ErrorOutputPaths:  []string{"stderr"},
+		DisableCaller:     true,
+		DisableStacktrace: true,
+		EncoderConfig: zapcore.EncoderConfig{
+			TimeKey:        "time",
+			LevelKey:       "level",
+			MessageKey:     "event",
+			EncodeLevel:    zapcore.LowercaseColorLevelEncoder,
+			EncodeTime:     zapcore.ISO8601TimeEncoder,
+			EncodeDuration: zapcore.StringDurationEncoder,
+		},
+	}
+)
+
+type Logger struct {
+	ServiceName    string
+	HasProbability bool
+	Probability    float64
+	Output         *zap.Logger
+}
+
+// New return a new Logger instance
+// add probability if the log shouldn't be logged always
+// the probability should be 1 based. e.g: 0.1 -> 10%
+func New(service string, hasProbability bool, probability float64, config *zap.Config) (*Logger, error) {
+	if service == "" {
+		panic(ErrEmptyService)
```

Comment:
I think this should return nil, err and be handled in the service where this is called rather than calling panic directly.

Reply:
🤔🤔 mmm Yea we can do that